### ISSUE TITLE
BN_mod_exp2_mont NULL pointer dereference if modulus is 0

### ISSUE THREAD

**Issue by guidovranken** [openssl/openssl#17648]
`BN_mod_exp2_mont` accesses the data field of the modulus without checking if it is allocated:

https://github.com/openssl/openssl/blob/77f3936928068bee9d7e0c6939709ac179cb1059/crypto/bn/bn_exp2.c#L35

So calling `BN_mod_exp2_mont` with a zero modulus can lead to a NULL pointer dereference.

Internally, this function is only used by DSA verification:

https://github.com/openssl/openssl/blob/77f3936928068bee9d7e0c6939709ac179cb1059/crypto/dsa/dsa_ossl.c#L402

It would be possible to crash a TLS client connection with crafted data (specifically by defining a zero `p` value) if it weren't for the fact that `dsa->flags` is by default set to `DSA_FLAG_CACHE_MONT_P`:

https://github.com/openssl/openssl/blob/77f3936928068bee9d7e0c6939709ac179cb1059/crypto/dsa/dsa_ossl.c#L428

This preloads a Montgomery structure with the modulus: https://github.com/openssl/openssl/blob/77f3936928068bee9d7e0c6939709ac179cb1059/crypto/dsa/dsa_ossl.c#L391

This rejects a zero modulus, so in that case `BN_mod_exp2_mont` is never reached, and a crash is avoided.

Apart from causing a NULL pointer dereference, this bug also might be leveraged as a side channel leak of 1 bit if the modulus passed to `BN_mod_exp2_mont` was previously used to store other (secret) data; data in `d` can be retained even if `top` is set to zero, for example after calling `BN_zero()`, and observing whether `BN_mod_exp2_mont` succeeds or fails reveals the least significant bit.

Despite the bug not being remotely reachable at this moment, I suggest adding a check to `BN_mod_exp2_mont` anyway:

- Users may be using `BN_mod_exp2_mont` directly
- Users can unset the `DSA_FLAG_CACHE_MONT_P` manually

For demonstration purposes, below is an input to the client fuzzer which causes it to crash if you disable `DSA_FLAG_CACHE_MONT_P`.

```
unsigned char crasher_bin[] = {
  0x16, 0x03, 0x0b, 0x00, 0x3d, 0x02, 0x00, 0x00, 0x39, 0x03, 0x03, 0x07,
  0x20, 0x77, 0x5d, 0xab, 0xdb, 0x08, 0x71, 0x9c, 0x3e, 0xe4, 0xb1, 0xf9,
  0xbc, 0xc2, 0xcf, 0xfc, 0x8d, 0xa0, 0x3f, 0x29, 0xd7, 0x66, 0xb7, 0x96,
  0xe7, 0x68, 0xfa, 0xa0, 0x6d, 0xf6, 0x0c, 0x00, 0x00, 0x32, 0x00, 0x00,
  0x11, 0xff, 0x01, 0x00, 0x01, 0x00, 0x32, 0x23, 0x00, 0x00, 0x00, 0x16,
  0x00, 0x00, 0x02, 0x17, 0x00, 0x00, 0x16, 0x03, 0x03, 0x04, 0x17, 0x0b,
  0x00, 0x04, 0x13, 0x00, 0x04, 0x10, 0x00, 0x04, 0x0d, 0x30, 0x82, 0x04,
  0x09, 0x30, 0x82, 0x02, 0xf1, 0xa0, 0x03, 0x02, 0x01, 0x02, 0x02, 0x14,
  0xa0, 0xae, 0x1e, 0x00, 0x08, 0x81, 0x0e, 0x44, 0x42, 0xff, 0xff, 0x00,
  0xcd, 0xc6, 0x45, 0xd8, 0x0c, 0x6d, 0xee, 0xd2, 0x30, 0x0d, 0x06, 0x09,
  0x2a, 0x86, 0x48, 0x86, 0xf7, 0xc1, 0xd9, 0x01, 0x05, 0x05, 0x00, 0x30,
  0x39, 0x31, 0x0b, 0x30, 0x09, 0x06, 0x03, 0x55, 0x03, 0x00, 0x13, 0x02,
  0x41, 0x55, 0x31, 0x17, 0x30, 0x15, 0x06, 0x03, 0x55, 0x31, 0x0a, 0x0c,
  0x0e, 0x44, 0x6f, 0x64, 0x67, 0x78, 0x67, 0x42, 0x72, 0x6f, 0x30, 0x68,
  0x0c, 0x72, 0x73, 0x31, 0x11, 0x30, 0x0f, 0x06, 0x03, 0xe0, 0x04, 0x03,
  0x0c, 0x08, 0x79, 0x67, 0x44, 0x64, 0x6f, 0x20, 0x43, 0x41, 0x30, 0x1e,
  0x17, 0x0d, 0x27, 0x39, 0x30, 0x68, 0x69, 0x6e, 0x64, 0x65, 0x78, 0x35,
  0x71, 0x37, 0x5a, 0x17, 0x0d, 0x31, 0x38, 0x30, 0x38, 0x2f, 0x30, 0x36,
  0x37, 0x33, 0x45, 0x31, 0x30, 0x5a, 0x30, 0x54, 0x31, 0x0b, 0x30, 0x09,
  0x06, 0x03, 0x12, 0x00, 0x2d, 0x13, 0x02, 0x41, 0x55, 0x31, 0x17, 0x30,
  0x15, 0x06, 0x03, 0x55, 0x04, 0x1a, 0x0c, 0x0e, 0x44, 0x6f, 0x64, 0x62,
  0x67, 0x6f, 0x79, 0x42, 0x20, 0x74, 0x68, 0x65, 0x72, 0x73, 0x31, 0x12,
  0x30, 0x10, 0x06, 0x03, 0x55, 0x04, 0x03, 0x0c, 0x09, 0x42, 0x72, 0x6b,
  0x74, 0x68, 0x65, 0x72, 0x20, 0x32, 0x31, 0x18, 0x30, 0x16, 0x06, 0x03,
  0x55, 0x04, 0x03, 0x0c, 0x0f, 0x52, 0x2c, 0x20, 0x53, 0x74, 0x43, 0x2a,
  0x66, 0x69, 0x69, 0x72, 0x63, 0x61, 0x11, 0x00, 0x30, 0x82, 0x01, 0xb6,
  0x30, 0x82, 0x01, 0x2b, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x38, 0x04,
  0x01, 0x30, 0x82, 0x01, 0x1e, 0x02, 0x81, 0x81, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x15, 0x00, 0xbb, 0x46, 0x5b, 0x34,
  0xff, 0xff, 0xfb, 0x41, 0x00, 0xda, 0x01, 0xff, 0xd3, 0x00, 0x02, 0x38,
  0x36, 0xdf, 0x6f, 0x75, 0x02, 0x81, 0x80, 0x90, 0xdd, 0x75, 0x59, 0x00,
  0x80, 0x00, 0x00, 0x95, 0x00, 0x8a, 0x5d, 0x2e, 0x2e, 0xea, 0x68, 0x6f,
  0x35, 0x57, 0x5a, 0x20, 0x00, 0x01, 0x00, 0x00, 0xee, 0xeb, 0x31, 0x0d,
  0x06, 0x09, 0x2a, 0xab, 0x3e, 0x06, 0xc3, 0x27, 0x6d, 0xff, 0xff, 0xff,
  0x05, 0x02, 0x00, 0x2e, 0x11, 0x3d, 0x5c, 0x18, 0x9d, 0x5b, 0x6b, 0x05,
  0x02, 0x00, 0xff, 0xff, 0xff, 0xbf, 0x9e, 0x07, 0xa8, 0xff, 0xff, 0xff,
  0x70, 0x2b, 0xa3, 0xe2, 0x00, 0xff, 0xff, 0x7a, 0xee, 0x75, 0x30, 0x8a,
  0xf9, 0x20, 0x00, 0xd0, 0x23, 0xe7, 0xd5, 0xe5, 0x9f, 0x3f, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x4b, 0xee, 0x4a, 0x8e, 0x26, 0xff, 0xff,
  0xff, 0x31, 0x02, 0x5a, 0xb3, 0xdb, 0x00, 0xe2, 0xe4, 0xff, 0xe4, 0xc0,
  0xfd, 0x13, 0xfb, 0xf5, 0xb5, 0x73, 0x85, 0x4a, 0x86, 0xdc, 0xff, 0xfd,
  0x54, 0x25, 0x4c, 0x03, 0x81, 0x84, 0x00, 0x02, 0x7d, 0x4c, 0x82, 0x01,
  0x1e, 0x42, 0x81, 0x06, 0x59, 0x81, 0x9e, 0xff, 0xfe, 0xff, 0x0d, 0x8f,
  0xd7, 0x04, 0x00, 0x00, 0x00, 0x51, 0x00, 0x10, 0x88, 0xf8, 0x16, 0x03,
  0x72, 0x00, 0x04, 0x02, 0x7a, 0xf5, 0xff, 0x33, 0x2e, 0x33, 0x2e, 0x95,
  0x30, 0x31, 0x2e, 0x31, 0x31, 0x30, 0x79, 0x70, 0x74, 0x65, 0x64, 0x70,
  0x72, 0x69, 0x75, 0x61, 0x74, 0x65, 0x6b, 0x9b, 0x00, 0xfd, 0x86, 0x00,
  0x00, 0x86, 0x00, 0x2f, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x02, 0x32, 0x37, 0x88, 0x9f, 0x99, 0xaf, 0xcd, 0x77, 0xc9, 0x09,
  0x13, 0xc7, 0x2d, 0x66, 0x52, 0x38, 0x2e, 0x4c, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x35, 0x95, 0x60, 0xff, 0x5c, 0xbb, 0x2d, 0x00, 0x00,
  0xff, 0x49, 0xb8, 0x77, 0x4c, 0x41, 0x55, 0xfe, 0x6b, 0x6e, 0x43, 0x4e,
  0xfe, 0x07, 0xff, 0xf9, 0x01, 0x95, 0xa3, 0x5a, 0x30, 0x58, 0x30, 0x1d,
  0x06, 0x03, 0x55, 0x5d, 0x16, 0x04, 0x16, 0x2c, 0xeb, 0x28, 0xad, 0xe0,
  0xfb, 0xab, 0xed, 0xf7, 0xc1, 0x2a, 0x7e, 0x2d, 0x9d, 0x93, 0x00, 0x00,
  0xc8, 0x0b, 0x96, 0x0a, 0xfa, 0x30, 0x1f, 0x06, 0x03, 0x54, 0x1d, 0x23,
  0x04, 0x18, 0x10, 0xc0, 0x19, 0x1e, 0x33, 0x20, 0x9c, 0xe9, 0x45, 0x68,
  0x71, 0x75, 0x21, 0x00, 0x80, 0x00, 0x00, 0x16, 0xde, 0x00, 0x02, 0x00,
  0x00, 0x71, 0x30, 0x09, 0x06, 0x03, 0x5c, 0x1d, 0x13, 0x04, 0x02, 0x35,
  0x7a, 0x30, 0x0b, 0x06, 0x03, 0x55, 0x11, 0x0f, 0x04, 0x04, 0xf5, 0xfd,
  0xf9, 0x82, 0x30, 0x0d, 0x06, 0x09, 0xcc, 0x79, 0x48, 0xdc, 0x02, 0xa0,
  0x36, 0x01, 0x05, 0xc2, 0x00, 0x03, 0x82, 0x01, 0x01, 0x00, 0x33, 0x99,
  0x30, 0x4e, 0x0e, 0xfa, 0x79, 0xc3, 0xbd, 0xcd, 0xc3, 0xde, 0xe1, 0xe9,
  0xa3, 0x9c, 0x65, 0x4e, 0x55, 0xbc, 0x4c, 0x20, 0x43, 0x65, 0x72, 0x32,
  0x35, 0x31, 0x5a, 0x17, 0x0d, 0xb0, 0x30, 0x7a, 0x39, 0x32, 0x10, 0x00,
  0x00, 0x00, 0x00, 0x00, 0xcd, 0x01, 0x30, 0x11, 0xcf, 0xf1, 0x30, 0x0d,
  0x30, 0x82, 0x01, 0x1e, 0x02, 0x81, 0x81, 0x42, 0x06, 0x5d, 0xff, 0xff,
  0xff, 0x07, 0x8f, 0x29, 0x7d, 0xf0, 0x9e, 0xf7, 0xb5, 0xff, 0x05, 0x67,
  0x07, 0x38, 0xf7, 0x01, 0x00, 0x09, 0x40, 0x00, 0x00, 0x40, 0x06, 0xc0,
  0x52, 0xb0, 0x2e, 0x0e, 0x00, 0x1f, 0x04, 0x07, 0x80, 0xc8, 0x8b, 0xf4,
  0x6d, 0x7a, 0x52, 0xff, 0xff, 0x03, 0xef, 0x00, 0x00, 0x00, 0xc8, 0x00,
  0x00, 0x04, 0x00, 0x40, 0x05, 0x04, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff,
  0xf6, 0xff, 0x32, 0x51, 0x65, 0x1d, 0x37, 0x88, 0x8f, 0x99, 0x00, 0x00,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x08, 0xff, 0x00, 0x00, 0x04, 0x00,
  0x40, 0x2c, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0a, 0x08, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x23, 0x33, 0x2e, 0x33, 0x2e, 0x30, 0x30, 0x31, 0x2e,
  0x31, 0x31, 0x00, 0x00, 0x00, 0x2d, 0x00, 0x00, 0x00, 0x01, 0x5b, 0x00,
  0x00, 0x00, 0x00, 0x49, 0x02, 0x1f, 0x00, 0xf7, 0x4a, 0x70, 0x53, 0x71,
  0x74, 0x32, 0x9c, 0x1b, 0x53, 0x10, 0xb3, 0x14, 0xcf, 0x15, 0xbb, 0x3d,
  0xcd, 0xe0, 0xcf, 0x21, 0xad, 0x74, 0xe5, 0x9a, 0x60, 0x11, 0xbe, 0x1d,
  0x8c, 0x02, 0x1e, 0x65, 0xb8, 0x91, 0xc2, 0xa2, 0x11, 0x16, 0x7a, 0xbb,
  0x2c, 0x5e, 0x07, 0x9e, 0x09, 0xe2, 0xc8, 0xa8, 0x34, 0x9c, 0x40, 0x00,
  0x00, 0xfc, 0xc5, 0x29, 0x4c, 0xed, 0xb4, 0x23, 0xaa, 0x11, 0x47, 0x7b,
  0x98, 0xc3, 0x16, 0x03, 0x03, 0x01, 0x3d, 0x0c, 0x00, 0x01, 0x39, 0x00,
  0x80, 0xf9, 0x3d, 0x16, 0x05, 0x00, 0x41, 0x26, 0xc9, 0xda, 0x3a, 0x70,
  0x18, 0x95, 0x60, 0x95, 0xc7, 0x00, 0x01, 0x00, 0x00, 0x56, 0x10, 0xb1,
  0x01, 0x00, 0x62, 0x00, 0x00, 0x00, 0x00, 0xb5, 0x70, 0x88, 0x78, 0xcd,
  0x43, 0x1f, 0x11, 0x16, 0x70, 0x91, 0x5b, 0xd2, 0x00, 0x09, 0x00, 0x58,
  0xb5, 0x4e, 0x44, 0xf3, 0x55, 0x04, 0x2a, 0x26, 0xf8, 0xfd, 0x34, 0x00,
  0x00, 0xff, 0xff, 0x0a, 0x00, 0x03, 0xa5, 0x41, 0xa6, 0x60, 0xdc, 0x53,
  0x00, 0x01, 0x02, 0x00, 0xb1, 0x01, 0x00, 0x00, 0x00, 0x00, 0xe1, 0x00,
  0x4f, 0xfd, 0x23, 0x43, 0x1f, 0xb5, 0x6f, 0x88, 0xcd, 0xa4, 0xed, 0x11,
  0x16, 0x71, 0x95, 0x5b, 0xd2, 0x00, 0x00, 0x00, 0x58, 0xb5, 0x4e, 0x44,
  0x71, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff,
  0xff, 0x00, 0x03, 0xa1, 0x41, 0xa6, 0x55, 0xdc, 0x53, 0x00, 0x01, 0x02,
  0x00, 0x80, 0x41, 0xc4, 0x04, 0x00, 0xc0, 0x08, 0x02, 0x00, 0xeb, 0x60,
  0x73, 0x28, 0x55, 0x10, 0x00, 0x00, 0x00, 0x00, 0x7f, 0xff, 0xff, 0xff,
  0x09, 0xde, 0xdc, 0x48, 0xe9, 0xfc, 0xca, 0x7d, 0xcc, 0x01, 0x05, 0x00,
  0x79, 0xc2, 0x01, 0x01, 0x29, 0x99, 0x30, 0x00, 0x26, 0x0e, 0x79, 0xff,
  0xfa, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0x01, 0x09, 0x00, 0x01, 0x04,
  0x00, 0x03, 0xcd, 0xa4, 0xed, 0x11, 0x16, 0x70, 0x95, 0x5b, 0xd2, 0x00,
  0x00, 0x00, 0x58, 0xb5, 0x4e, 0x44, 0x71, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x20, 0x04, 0x00, 0x00, 0x00, 0xff, 0x00, 0x43, 0xa1, 0x41, 0xa6,
  0x55, 0xdc, 0x53, 0x00, 0x01, 0x02, 0x00, 0x80, 0x41, 0x3f, 0x04, 0x00,
  0xc0, 0x08, 0x02, 0x00, 0xeb, 0x6d, 0xee, 0xeb, 0x31, 0x0d, 0x06, 0x09,
  0x2a, 0xab, 0x3e, 0x06, 0xc3, 0x27, 0x6d, 0xff, 0xff, 0xff, 0x05, 0x02,
  0x00, 0x2e, 0x30, 0x2c, 0x02, 0x14, 0x26, 0xe1, 0x00, 0xcc, 0xdc, 0x79,
  0x48, 0x3d, 0x02, 0x01, 0x06, 0x48, 0x2a, 0x48, 0x2d, 0xce, 0x09, 0xd1,
  0xcc, 0x7e, 0x02, 0x14, 0x33, 0x63, 0x2a, 0x86, 0x73, 0x72, 0x69, 0x6e,
  0x63, 0x75, 0x32, 0x79, 0x74, 0xbb, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};
unsigned int crasher_bin_len = 1440;
```

**Comment by guidovranken** [comment]
LibreSSL fix: https://github.com/libressl-portable/openbsd/commit/f4d5104dbf792556b23da0cb184f75d503d0a063

**Comment by hlandau** [comment]
Fixed by #17783.
